docs/code-standards-policy.rst

```diff
@@ -118,36 +171,418 @@ using the debops.apt_install_ role MUST be defined as (or similar):
+via Sphinx_.
+
+Each variable comment is started with a ``.. envvar::`` reference anchor
+followed by the name of the variable. This construct allows any documentation
+page to reference the variable via ``:envvar:`varname``` which :program:`Sphinx`
+will translate into a link pointing to the variable description. Below the
+anchor a comment should describe the purpose of the variable, accepted values,
+side effects and so on. Within the comment all reStructuredText constructs
+supported by :program:`Sphinx` can be used. An example variable definition would
+eventually look like this:
+
+.. code-block:: yaml
+
+    # .. envvar:: ferm__flush [[[
+    #
+    # Should ferm-rules be flushed when :program:`ferm` is disabled? The default is true,
+    # but you may need set both :envvar:`ferm__enabled` and this to ``False`` if you are
+    # running in some container and are not allowed to change :command:`iptables`.
+    ferm__flush: '{{ ferm__enabled | bool }}'
+                                                                   # ]]]
+
+Related default variables should be grouped to sections for a better overview
+and easier navigation. For example it makes sense to distinguish packaging and
+network related variables. Of course additional or different section titles
+might be meaningful for the individual role:
+
+.. code-block:: yaml
+
+    # APT packages [[[
+    # ----------------
+
+    [... packaging related variables ...]
+
+                                                                   # ]]]
+    # Network configuration [[[
+    # -------------------------
+
+    [... networking related variables ...]
+                                                                   # ]]]
+
+.. _debops_policy__ref_code_standards_dependent_variables:
+
+Dependent variables
+-------------------
+
+DebOps is designed in a way that there are divided responsibilities between the
+roles. Each role has its clear task to fulfill. For example an application role
+must never care about the firewall configuration by itself. There is a dedicated
+role for firewall configuration. And the application role needs a way to tell
+the firewall role which access rules to configure. This is done via dependent
+variables.
+
+Each role providing a feature which MAY be consumed by other roles MUST provide
+a dedicated dependent variable for the configuration of this feature. As that
+variable is always defined in the context of the consuming role, its value MUST
+NOT modify a shared configuration state (e.g. configuration template). The
+provider role might be executed multiple times depending on the role dependency
+configuration of the consuming roles.
+
+.. caution::
+
+   A role providing dependent variables MUST be able to handle multiple
+   role executions with different values of the dependent variables without
+   mutual interference.
+
+Ideally the dependent variable SHOULD accept a list of YAML dictionaries
```

Comment:
This can be a YAML dictionary of YAML dictionaries as well, because in Ansible 2.0.0+ the `combine()` filter is available which can be used to merge different YAML dictionaries together. This was recently done in the `debops.ifupdown` role and I think a few more roles can get that treatment now. Advantage of using a merged dictionary is an easier way to modify existing entries without adding new ones. Making it more user-friendly is a bit complicated since you probably want to write a Jinja tempate to do custom merging, but it's certainly doable.